### PR TITLE
[TRTLLM-15498][refactor] consolidate Kimi KDA production frontend

### PR DESCRIPTION
**Pull request by jiaganc**
## Description

Consolidate the Kimi K3 KDA production frontend in `KimiKDALinearAttention` and remove the duplicate `KimiKDARuntime` wrapper from `modeling_kimi_linear.py`.

The previous implementation split KDA ownership across two classes: a standalone parity-oriented module and a separate executor-facing runtime. That duplicated projection, convolution, dispatch, and state-management logic and made it unclear which frontend was used in production.

This PR:

- moves the pool-backed production prefill, decode, and speculative-verify paths directly into `KimiKDALinearAttention`;
- keeps the Blackwell optimized paths in the production methods and makes the portable prefill/decode fallbacks explicit;
- preserves the checkpoint-facing module hierarchy and parameter names;
- moves the standalone parity façade to the test tree as `KimiKDAReference`;
- removes `KimiKDARuntime` and unused/test-only production helpers;
- retains small-varlen prefill fallback routing required by the Blackwell persistent scheduler.

The result gives Kimi KDA one production owner while keeping the reference implementation and parity utilities out of the runtime package.

## Test Coverage

- `pre-commit` on all modified files: passed.
- B200 focused regression `test_kimi_kda_verify_parity.py::test_kda_fused_prefill_matches_separate_projections`: `1 passed`.
- B200 focused KDA unit gate: `61 passed, 15 skipped, 0 failed`. The gate covered `tests/unittest/_torch/modules/kimi_kda` plus the KDA MTP, frontend, fused-verify, and verify-parity modeling tests.
- TensorRT-LLM was compiled once for SM100 and the resulting build was reused for the Python-only follow-up changes.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Dev Engineer Review

- Consolidates production prefill, decode, and speculative verification in `KimiKDALinearAttention`.
- Removes `KimiKDARuntime` and unused production APIs.
- Preserves checkpoint hierarchy, parameter names, sharding, and post-load handling.
- Retains optimized paths, portable fallbacks, and small-varlen routing.
- Adds optimized-prefill eligibility checks.
- Moves the parity façade to tests as `KimiKDAReference`.
- Updates pooled-state handling, metadata flow, FP8 conversion, and verification finalization.
- Reported validation passed: pre-commit, focused parity coverage, and 61 KDA unit tests. Fifteen tests were skipped.

## QA Engineer Review

- Adds `test_checkpoint_plan_preserves_external_attention_names`.
- Adds `KimiKDAReference` and test-only cache utilities.
- Adds `test_decode_fallback_matches_fla_reference`.
- Updates decode tests for optimized dispatch, fallback behavior, indexed state pools, head-count boundaries, and CUDA-graph safety.
- Updates prefill tests for fixed-length, empty, varlen, partial-chunk, dispatch-boundary, and FLA fallback cases.
- Updates state-parity, cache-soundness, fused-verification, speculative-verification, and MTP decode parity tests.
- Updates `tests/integration/test_lists/test-db/l0_b200.yml` and `tests/integration/test_lists/test-db/l0_gb300_multi_gpus.yml` to reference the KDA parity test modules.
- The modified test modules are covered by the updated `test-db/` selectors.
- CBTS coverage data is unavailable.
- Verdict: needs follow-up.

### CODE REVIEW

**Review by xinhe-nv — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_kernels.py [hunk 1]
```diff
@@ -236,6 +222,21 @@ def mtp_verify(self, **kwargs) -> torch.Tensor:
...
+        cu_seqlens: Optional[torch.Tensor],
+        chunk_size: int = 64,
+    ) -> bool:
+        """Return whether this batch can stay on the Blackwell prefill path."""
+        if self.prefill_kernel_path != "optimized":
+            return False
+        if cu_seqlens is not None:
+            from fla.ops.utils.index import prepare_chunk_indices
+
+            return prepare_chunk_indices(cu_seqlens, chunk_size).shape[0] >= 4
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/kimi_kda/_kda_kernels.py]: This ≥4-chunk rule now exists twice: here (frontend routing) and inline in `prefill_chunk_kda` (the `chunk_indices.shape[0] < 4` fallback around line 291), each with its own `chunk_size=64` default. If either the threshold or the chunk size is ever retuned in one place only, the frontend and the dispatch silently disagree — benign today (the inner check still prevents the zero-size-grid launch) but wasteful and confusing. Since the inner check must stay for callers that hit `prefill_chunk_kda` directly, have it call `can_use_optimized_prefill` so the predicate has a single owner. Also note both calls run `prepare_chunk_indices` (a host-syncing computation); dedup currently relies on FLA's tensor cache seeing the same `cu_seqlens` object — worth a comment either way.

File: tests/unittest/_torch/modules/kimi_kda/kimi_kda_test_utils.py [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+            kernel_size=conv_kernel_size,
+            activation="silu",
+        )
+
+        self.A_log = nn.Parameter(
+            torch.log(torch.empty(num_heads, dtype=torch.float32).uniform_(1, 16))
+        )
+        self.f_a_proj = nn.Linear(hidden_size, head_dim, bias=False)
+        self.f_b_proj = nn.Linear(head_dim, projection_size, bias=False)
+        self.dt_bias = nn.Parameter(torch.empty(projection_size, dtype=torch.float32))
```
Comment (brnguyen2) [inline tests/unittest/_torch/modules/kimi_kda/kimi_kda_test_utils.py]: `torch.empty` here reintroduces the hazard TRTLLM-15204 fixed: heap garbage can contain NaN bit patterns that poison both the optimized and FLA gates in randomly-constructed modules. The production module kept its `uniform_(log 1e-3, log 1e-1)` init (with the explanatory comment), but this façade lost it in the move. Every current test happens to overwrite dt_bias via `load_state_dict` from a production module or an explicit `normal_()`, but this is a shared utility — the first future test that constructs `KimiKDAReference` standalone gets nondeterministic NaN flakes. Keep the same uniform init here (MetaInitMode isn't a constraint in the test tree, so even a plain init works).

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py [hunk 3]
```diff
@@ -256,482 +162,1028 @@ def __init__(
...
-        cu_seqlens : optional cumulative sequence lengths for varlen inputs.
-
-        Returns
-        -------
-        ``(B, T, hidden_size)`` output tensor (equal-length case) or
-        ``(1, sum(seq_lens), hidden_size)`` (varlen case).
-        """
-        if cu_seqlens is not None:
-            cu_seqlens = cu_seqlens.to(device=hidden_states.device, dtype=torch.long)
+        if not self._dispatch.can_use_optimized_prefill(cu_seqlens=cu_seqlens):
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py]: This frontend routing is a behavior change relative to the removed `KimiKDARuntime._forward_prefill`, which always used the fused `_qkvg_proj_weight`/`qkvg_proj` projections and relied on the dispatch-internal FLA fallback for small-varlen batches. Now those batches — and every prefill when `TLLM_KDA_ENABLE_OPT_PREFILL=0` on Blackwell — run `forward_prefill_fallback`'s five separate projection GEMMs and skip the fused onorm gate output. Correct (the per-projection modules remain valid views after both finalize variants and the FP8 conversion) and negligible at small-varlen sizes, but the env-var opt-out path now de-fuses prefill projections too; if that opt-out is used for perf triage, the comparison is no longer apples-to-apples with the old runtime. Worth a line in the docstring or PR description stating this is intended.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The consolidation is faithful — I diffed the moved prefill/decode/verify/finalize bodies against the deleted `KimiKDARuntime` and they are near-verbatim, checkpoint names are covered by the new plan test, and the moved test files are fully re-registered (both l0 lists, the monkeypatch, no stale `-k` patterns or waives).

One behavior change is worth a sentence in the PR description: the old production path always ran the fused projections and let `prefill_chunk_kda` fall back to FLA internally for small-varlen batches; now `can_use_optimized_prefill` routes those batches (and the entire `TLLM_KDA_ENABLE_OPT_PREFILL=0` opt-out on Blackwell) into `forward_prefill_fallback`, which re-runs five separate projection GEMMs and never touches the fused weights. It's correct — the per-projection modules stay valid as views after both the BF16 merge and the FP8 conversion — and negligible at these batch sizes, but it's a routing change, not pure motion, and the new small-varlen parity tests are what make it safe.

Two small items inline: the ≥4-chunk eligibility rule now exists in two places that can drift, and the test-tree `KimiKDAReference` dropped the dt_bias initialization that TRTLLM-15204 added to the production module.

Minor observability note: `forward_decode` dropped the `info_once` lines that recorded whether the indexed or static recurrent-state path was active. The dispatch-selection log at init remains, but the indexed-pool choice is runtime-conditional (`ssm_state_indices` availability) and is now silent; fine if intentional.

File: tensorrt_llm/_torch/modules/kimi_kda/_kda_kernels.py (hunk 1 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/_kda_kernels.py]: Addressed in added5d493. I removed the frontend eligibility helper and the duplicated prefill fallback frontend. `forward_prefill()` now always runs the shared projection/convolution/cache path and calls `prefill_chunk_kda()`, which is the single owner of the optimized-versus-FLA core selection, including the <4-chunk guard and `prepare_chunk_indices` call. I also removed `TLLM_KDA_ENABLE_OPT_PREFILL`. The host-sync optimization is intentionally deferred to a later PR, and decode is unchanged in this follow-up.

File: tests/unittest/_torch/modules/kimi_kda/kimi_kda_test_utils.py (hunk 2 above)
Comment (jiaganc) [reply tests/unittest/_torch/modules/kimi_kda/kimi_kda_test_utils.py]: Fixed in e42e1b8a6d: `KimiKDAReference.dt_bias` now uses the same `uniform_(log(1e-3), log(1e-1))` initialization as production, so standalone construction cannot consume uninitialized NaN bit patterns. Follow-up 65ec6f8bdc also makes `KimiKDAReference` an FLA-only oracle: prefill calls `fla.ops.kda.chunk_kda` directly, decode calls `fla.ops.kda.fused_recurrent_kda` directly, and the optimized dispatcher/path helpers were removed from the reference. B200 validation passed: 23 passed, 14 skipped.

File: tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py (hunk 3 above)
Comment (jiaganc) [reply tensorrt_llm/_torch/modules/kimi_kda/kimi_kda_mixer.py]: Addressed in added5d493. `TLLM_KDA_ENABLE_OPT_PREFILL` and the duplicated `forward_prefill_fallback()` frontend were removed. `forward_prefill()` now always uses the common fused QKVG/BFA projection and cache path, then `prefill_chunk_kda()` alone selects the optimized or FLA core (including small-varlen fallback). Therefore the FLA core fallback no longer de-fuses projections, and there is no environment opt-out path with different projection behavior. The updated B200 parity gate passed: 23 passed, 14 skipped.

**Review by xxi-nv — APPROVED** [review]

**Review by lori-ren — APPROVED** [review]

**Review by BowenFu — APPROVED** [review]

**Review by sunnyqgg — APPROVED** [review]
LGTM for moe part

**Review by yingguo-trt — APPROVED** [review]

**Review by zhaoyuanh-nvidia — APPROVED** [review]

**Review by chenfeiz0326 — APPROVED** [review]
